src/main/scala/co/teapot/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -15,51 +15,77 @@
+import co.teapot.graph.EdgeDir.EdgeDir
+import co.teapot.io.ByteBufferIntSlice
 
-import scala.collection.mutable
-import scala.collection.mutable.ArrayBuffer
-import scala.io.Source
+// Make the constants in MemoryMappedDirectedGraph available
+import co.teapot.graph.MemoryMappedDirectedGraph._
 
 /**
  * A graph which reads edge data from a memory mapped file.  There is no object overhead per
- * node: the memory used for n nodes and m edges with both in-neighbor and out-neighbor access is
- * exactly 16 + 16*n + 8*m bytes. Also, loading is very fast because no parsing of text is required.
+ * node: the memory used for with both in-neighbor and out-neighbor access is 8 bytes per edge
+  * and 8 bytes per node. Loading is very fast because no parsing of text is required.
  * Loading time is exactly the time it takes the operating system to page data from disk into
  * memory.  Nodes are numbered sequentially from 0 to nodeCount - 1 and must be a range of this
  * form (i.e. nodeCount == maxNodeId + 1).
  *
- * When transforming a graph where nodeCount <= maxNodeId
+ * Graphs are converted to this format using MemoryMappedDirectedGraphConverter, then loaded
+  * using this class.  When transforming a graph where nodeCount <= maxNodeId
  * to this format, new nodes with no neighbors will be implicitly created. The binary
  * format is currently subject to change.
  */
 
-/* Storage format
+/* Internally, the graph is stored in a large binary file made of a header and a series of segments.
+Out-neighbor segment i contains the out-neighbors of all nodes whose id satisfies
+  id % segmentCount == i
+and similarly for in-neighbor segment i.  Each segment has a sequence of offsets, where the jth
+offset in out-neighbor segment i stores the offset (as an int relative to the start of that segment) for
+node
+  id = j * segmentCount + i
+where nodesPerSegment = nodeCount / segmentCount.  We require for simplicity that nodeCount be
+a multiple of segmentCount (by creating empty nodes if needed).
+
+Storage format
+Header:
 byteCount  data
-8          (reserved, later use for versioning or indicating undirected vs directed)
-8          n (i. e. the number of nodes).  Currently must be less than 2^31.
-8*(n+1)    Offsets into out-neighbor data. Index i (a Long) points to the out-neighbor data of
-           node i.  The out-neighbor data must be stored in sequential order by id, as the
-           outegree of node i is computed from the difference in offset between node i+1 and node i.
-           Index n is needed to compute the outdegree of node n - 1.
-8*(n+1)    Offsets into in-neighbor data (Longs) (Same interpretation as out-neighbor offsets)
-4*m        out-neighbor data
-4*m        in-neighbor data
+8          (version number, currently 1)
+4          n (nodeCount).  Must be a multiple of segmentCount.
+8          edgeCount
+4          k (segmentCount)
+8*(k + 1)  The absolute offset in bytes where each out-neighbor segment's data begins.  The "+ 1"
```

Comment:
So is the hope here that this would be in cache and hence the two memory accesses to get an edge would not hurt, right?